app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +71,90 @@ def command_prefix
   end
 
   def logger
     Rails.logger
   end
 
-  def initiate_connection!
-    Timeout.timeout(Setting[:ssh_timeout].to_i) do
-      Timeout.timeout(8) do
-        ssh.run('pwd')
-      end
-    rescue Errno::ECONNREFUSED
-      logger.debug "Connection refused for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Errno::EHOSTUNREACH
-      logger.debug "Host unreachable for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def scp
+    logger.debug "Uploading template to remote system '#{remote_host}'"
+    logger.debug scp_cmd
+    _stdout, stderr, status = Open3.capture3(*scp_cmd)
+
+    return if status.success?
+
+    logger.error stderr
+    raise Foreman::Exception.new("Failed to upload '#{@template}' to remote system at #{address}")
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def passwd_cmd(mask_passwd: false)
+    return if @options[:password].empty?
+
+    value = mask_passwd ? '*****' : @options[:password]
+    ["sshpass", "-p", value]
   end
 
-  def scp
-    Fog::SCP.new(address, username, options)
+  def ssh_key
+    return if @options[:key_data].empty?
+    ["-i", @tmp_key_path]
+  end
+
+  def ssh_cmd(host_command, mask_passwd: false)
+    [passwd_cmd, "ssh", ssh_key, ssh_options, remote_host, host_command].flatten.compact
+  end
+
+  def scp_cmd
+    [passwd_cmd, "scp", ssh_key, ssh_options, @template, "#{remote_host}:~/#{remote_script}"].flatten.compact
+  end
+
+  def ssh_options
+    opts = []
+
+    opts << "-o ConnectTimeout=6"
+    opts << "-o Compression=yes"
+    opts << "-o ControlMaster=auto"
+    opts << "-o ControlPath=#{socket_file}"
+    opts << "-o ControlPersist=yes"
+
+    methods = []
+    methods << 'publickey' if @options[:key_data].present?
+    methods << 'password' if @options[:password].present?
+
+    opts << "-o PreferredAuthentications=#{methods.join(',')}"
+
+    opts
+  end
+
+  def create_tmp_key
+    return if @options[:key_data].empty?
+
+    file = Tempfile.new("ssh-key-#{uuid}")
+    file.write(@options[:key_data])
+    file.close
+    file.path
+  end
+
+  def check_auth_options
+    return if @options[:key_data].present? || @options[:password].present?
+
+    raise Foreman::Exception.new("Missing :key_data or :password options, no authentication method available")
+  end
+
+  def socket_file
+    file = Tempfile.new("ssh-socket-#{@uuid}")
+    file.close
+    file.path
+  end
```

Comment:
This will return a new file on every invocation. Perhaps this is what you intended?
```suggestion
  def socket_file
    @socket_file ||= begin
      Tempfile.new("ssh-socket-#{@uuid}")
      file.close
      file.path
    end
  end
```

An alternative is to enhance `deploy!` to use block form and automatically take care of the cleanup in one go. You'd pass the socket file to ssh and scp, but that also gives you better options to test those specific methods.
```ruby
def deploy!
  Tempfile.create do |socket_file|
    ssh(socket_file.path)
    scp(socket_file.path)
  end
end
```